predicators/approaches/delivery_approach.py

```diff
@@ -0,0 +1,197 @@
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # Extracts domain-specific class instances for the delivery domain
+        types = {t.name: t for t in self._types}
+        self.type_loc = types["loc"]
+        self.type_paper = types["paper"]
+
+        options = {o.name: o for o in self._initial_options}
+        self.opt_pick_up = options["pick-up"]
+        self.opt_move = options["move"]
+        self.opt_deliver = options["deliver"]
+
+        predicates = {p.name: p for p in self._initial_predicates}
+        self.pred_is_home_base = predicates["ishomebase"]
+        self.pred_safe = predicates["safe"]
+        self.pred_at = predicates["at"]
+        self.pred_wants_paper = predicates["wantspaper"]
+        self.pred_satisfied = predicates["satisfied"]
+        self.pred_unpacked = predicates["unpacked"]
+        self.pred_carrying = predicates["carrying"]
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "delivery_policy"
+
+    @property
+    def is_learning_based(self) -> bool:
+        return False
+
+    def _extract_at_loc_optional(
+            self, locs: Iterable[Object],
+            ground_atoms: Iterable[GroundAtom]) -> Optional[Object]:
+        at_locs = filter_predicated_objs(locs, self.pred_at, ground_atoms)
+        if len(at_locs) > 1:
+            raise ApproachFailure(
+                "Cannot be at multiple locations at the same time")
+        return next(iter(at_locs), None)
+
+    def _extract_at_loc(self, locs: Iterable[Object],
+                        ground_atoms: Iterable[GroundAtom]) -> Object:
+        at_loc = self._extract_at_loc_optional(locs, ground_atoms)
+        if at_loc is None:
+            raise ApproachFailure("Expected a concrete location")
+        return at_loc
+
+    def _move(self, state: _PDDLEnvState, from_loc: Object,
+              to_loc: Object) -> Action:
+        ground_option = self.opt_move.ground([from_loc, to_loc],
+                                             np.empty(0, dtype=np.float32))
+        check_option(ground_option, state)
+        return ground_option.policy(state)
+
+    def _pick_up(self, state: _PDDLEnvState, paper: Object,
+                 loc: Object) -> Action:
+        ground_option = self.opt_pick_up.ground([paper, loc],
+                                                np.empty(0, dtype=np.float32))
+        check_option(ground_option, state)
+        return ground_option.policy(state)
+
+    def _deliver(self, state: _PDDLEnvState, paper: Object,
+                 loc: Object) -> Action:
+        ground_option = self.opt_deliver.ground([paper, loc],
+                                                np.empty(0, dtype=np.float32))
+        check_option(ground_option, state)
+        return ground_option.policy(state)
+
+    def _solve(self, task: Task, timeout: int) -> Callable[[State], Action]:
+        # Extracting task-specific information
+        init_state = cast(_PDDLEnvState, task.init)
```

Comment:
slightly better than using `cast` here is `assert isinstance(_PDDLEnvState, task.init)` which should satisfy mypy and has the added benefit that the code will actually crash (at runtime) if the assumption is violated